sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Call.java

```diff
@@ -17,75 +17,271 @@
+  }
+
+  /**
+   * Overrides the default {@link #DEFAULT_TIMEOUT}. A {@link UserCodeTimeoutException} is thrown
+   * when {@link Caller#call}, {@link SetupTeardown#setup}, or {@link SetupTeardown#teardown} exceed
+   * the timeout.
+   */
+  Call<RequestT, ResponseT> withTimeout(Duration timeout) {
+    return new Call<>(configuration.toBuilder().setTimeout(timeout).build());
+  }
+
+  @Override
+  public @NonNull Result<ResponseT> expand(PCollection<RequestT> input) {
+    TupleTag<ResponseT> responseTag = new TupleTag<ResponseT>() {};
+
+    PCollectionTuple pct =
+        input.apply(
+            CallFn.class.getSimpleName(),
+            ParDo.of(new CallFn<>(responseTag, configuration))
+                .withOutputTags(responseTag, TupleTagList.of(FAILURE_TAG)));
+
+    return Result.of(configuration.getResponseCoder(), responseTag, pct);
+  }
+
+  private static class CallFn<RequestT, ResponseT> extends DoFn<RequestT, ResponseT> {
+    private final TupleTag<ResponseT> responseTag;
+    private final CallerWithTimeout<RequestT, ResponseT> caller;
+    private final SetupTeardownWithTimeout setupTeardown;
+
+    private transient @MonotonicNonNull ExecutorService executor;
+
+    private CallFn(
+        TupleTag<ResponseT> responseTag, Configuration<RequestT, ResponseT> configuration) {
+      this.responseTag = responseTag;
+      this.caller = new CallerWithTimeout<>(configuration.getTimeout(), configuration.getCaller());
+      this.setupTeardown =
+          new SetupTeardownWithTimeout(
+              configuration.getTimeout(), configuration.getSetupTeardown());
+    }
+
+    /**
+     * Invokes {@link SetupTeardown#setup} forwarding its {@link UserCodeExecutionException}, if
+     * thrown.
+     */
+    @Setup
+    public void setup() throws UserCodeExecutionException {
+      this.executor = Executors.newSingleThreadExecutor();
+      this.caller.setExecutor(executor);
+      this.setupTeardown.setExecutor(executor);
+
+      // TODO(damondouglas): Incorporate repeater when https://github.com/apache/beam/issues/28926
+      //  resolves.
+      this.setupTeardown.setup();
+    }
+
+    /**
+     * Invokes {@link SetupTeardown#teardown} forwarding its {@link UserCodeExecutionException}, if
+     * thrown.
+     */
+    @Teardown
+    public void teardown() throws UserCodeExecutionException {
+      // TODO(damondouglas): Incorporate repeater when https://github.com/apache/beam/issues/28926
+      //  resolves.
+      this.setupTeardown.teardown();
+      checkStateNotNull(executor).shutdown();
+      try {
+        boolean ignored = executor.awaitTermination(3L, TimeUnit.SECONDS);
+      } catch (InterruptedException ignored) {
+      }
+    }
+
+    @ProcessElement
+    public void process(@Element @NonNull RequestT request, MultiOutputReceiver receiver)
+        throws JsonProcessingException {
+      try {
+        ResponseT response = this.caller.call(request);
```

Comment:
Let's also add a TODO here that mentions https://github.com/apache/beam/issues/29248.